unic/char/property/src/macros.rs

```diff
@@ -108,63 +147,229 @@ macro_rules! char_property {
+            fn bool(&self) -> bool { self.$bool_name() }
+        }
+
+        impl From<bool> for $name {
+            fn from(b: bool) -> Self { $name(b) }
+        }
+
+        impl From<$name> for bool {
+            fn from(prop: $name) -> bool { prop.$bool_name() }
+        }
+
+        char_property! {
+            __impl FromStr for $name;
+            // Yes
+            y => $name(true);
+            yes => $name(true);
+            t => $name(true);
+            true => $name(true);
+            // No
+            n => $name(false);
+            no => $name(false);
+            f => $name(false);
+            false => $name(false);
+        }
+
+        char_property! {
+            __impl Display for $name by BinaryCharProperty
+        }
+
+        char_property! {
+            __impl CharTraitBinary for $name;
+            $bool_name;
+            $(#[$char_trait_meta])* $char_trait_name;
+            $(#[$char_trait_fn_meta])* $char_trait_fn_name;
+        }
+
+        char_property! {
+            __impl StrTraitBinary for $name;
+            $bool_name;
+            $(#[$str_trait_meta])* $str_trait_name;
+            $(#[$str_trait_fn_meta])* $str_trait_fn_name;
+        }
+    };
+
+    // == Shared == //
+
+    ( __impl CharProperty for $name:ident;
+        $abbr:expr;
+        $long:expr;
+        $human:expr;
+    ) => {
+        impl $crate::CharProperty for $name {
+            fn prop_abbr_name() -> &'static str { $abbr }
+            fn prop_long_name() -> &'static str { $long }
+            fn prop_human_name() -> &'static str { $human }
+        }
+    };
+
+    ( __impl FromStr for $name:ident;
+        $( $id:ident => $value:expr; )*
+    ) => {
+        #[allow(unreachable_patterns)]
+        impl ::std::str::FromStr for $name {
+            type Err = ();
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                match s {
+                    // This stringify! should be moved out of this block to the call site. See the
+                    // test failure https://travis-ci.org/behnam/rust-unic/builds/275758001 for why
+                    // this is done here. This can be reverted at 1.20 adoption time.
+                    $( stringify!($id) => Ok($value), )*
+                    $( s if ::std::ascii::AsciiExt::eq_ignore_ascii_case(s, stringify!($id))
+                         => Ok($value), )*
+                    _ => Err(()),
                 }
             }
         }
+    };
 
+    ( __impl Display for $name:ident by $trait:ident ) => {
         impl ::std::fmt::Display for $name {
             fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
-                $crate::EnumeratedCharProperty::human_name(self).fmt(f)
+                $crate::$trait::human_name(self).fmt(f)
+            }
+        }
+    };
+
+    ( __impl CharTraitBinary for $name:ident;
+        $bool_name:ident;
+        $(#[$char_trait_meta:meta])* $char_trait_name:ident;
```

Comment:
Here's the 1.17 failure.

```
  --> unic/ucd/bidi/src/bidi_control.rs:19:1
   |
19 | char_property! {
   | ^
   |
   = note: this error originates in a macro outside of the current crate
```

You can't have a repeating group terminated by an `:ident` token until 1.20. You'll need to add an explicit sigil between the meta and the ident, or move the meta after the ident.

I'll do a more full review later today, but I wanted to point this out quickly so you'd know what to fix.

Reply:
Thanks, @CAD97! Fixed in the a new commit.